### PR TITLE
log-tag feature

### PR DESCRIPTION
**Pull request by h8718**
## Description

Moving out logger to a separate pkg 
## Purpose

- [ ] Bug fix
- [x] New feature
- [x] Other

### CODE REVIEW

**Comment by h8718** [comment]
TODO: 
```
panic: reflect: call of reflect.Value.NumField on map Value

goroutine 53464 [running]:
reflect.flag.mustBe(...)
        /usr/local/go/src/reflect/value.go:221
reflect.Value.NumField({0x23c33c0?, 0xc00081c5a0?, 0xc000611b90?})
        /usr/local/go/src/reflect/value.go:1963 +0x8d
github.com/direktiv/direktiv/pkg/flow/bytedata.convertDataForOutputStructBuilder({0x2b89e10, 0x23c33c0}, {0x23c33c0?, 0xc00081c5a0?, 0x2516b40?})
        /go/src/pkg/flow/bytedata/data.go:83 +0xeb
github.com/direktiv/direktiv/pkg/flow/bytedata.convertDataForOutputBuilder({0x23c33c0?, 0xc00081c5a0?})
        /go/src/pkg/flow/bytedata/data.go:168 +0x2e6
github.com/direktiv/direktiv/pkg/flow/bytedata.convertDataForOutputStructBuilder({0x2b89e10, 0x2676900}, {0x2676900?, 0xc00099a2a0?, 0x7f98186fcf18?})
        /go/src/pkg/flow/bytedata/data.go:115 +0x3fa
github.com/direktiv/direktiv/pkg/flow/bytedata.convertDataForOutputBuilder({0x2614920?, 0xc00099a2a0?})
        /go/src/pkg/flow/bytedata/data.go:168 +0x2e6
github.com/direktiv/direktiv/pkg/flow/bytedata.convertDataForOutputSliceBuilder({0x7f97f100b5a8?, 0x6?}, {0x227a860?, 0xc0003a65e8?, 0x7f98186fcf18?})
        /go/src/pkg/flow/bytedata/data.go:136 +0x1fd
github.com/direktiv/direktiv/pkg/flow/bytedata.convertDataForOutputBuilder({0x227a860?, 0xc0003a65e8?})
        /go/src/pkg/flow/bytedata/data.go:157 +0x332
github.com/direktiv/direktiv/pkg/flow/bytedata.ConvertDataForOutput({0x227a860?, 0xc0003a65e8?}, {0x224c960, 0xc000039080})
        /go/src/pkg/flow/bytedata/data.go:181 +0x2d
github.com/direktiv/direktiv/pkg/flow.(*flow).InstanceLogsParcels(0xc00043c9c0, 0xc000e7a550, {0x2b73820, 0xc000eed2d0})
        /go/src/pkg/flow/grpc-logs.go:379 +0x98a
github.com/direktiv/direktiv/pkg/flow/grpc._Flow_InstanceLogsParcels_Handler({0x273c200?, 0xc00043c9c0}, {0x2b719c0, 0xc000efe5b0})
        /go/src/pkg/flow/grpc/protocol_grpc.pb.go:4187 +0xd3
github.com/direktiv/direktiv/pkg/flow.streamInterceptor({0x273c200?, 0xc00043c9c0?}, {0x2b719c0?, 0xc000efe5b0?}, 0x2265ae0?, 0xc000eed0d0?)
        /go/src/pkg/flow/server.go:477 +0x30
google.golang.org/grpc.(*Server).processStreamingRPC(0xc00064a3c0, {0x2b78c58, 0xc0008b24e0}, 0xc001020900, 0xc00043cb70, 0x40ee1a0, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1565 +0xfa6
google.golang.org/grpc.(*Server).handleStream(0xc00064a3c0, {0x2b78c58, 0xc0008b24e0}, 0xc001020900, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1640 +0x9ea
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        /go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:932 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:930 +0x28a
```

File: pkg/flow/engine.go [hunk 1]
```diff
@@ -396,14 +400,14 @@ func (engine *engine) CrashInstance(ctx context.Context, im *instanceMemory, err
 
 	if errors.As(err, &cerr) {
 		engine.sugar.Errorf("Instance failed with error '%s': %v", cerr.Code, err)
-		engine.logToInstance(ctx, time.Now(), im.cached, "Instance failed with error '%s': %s", cerr.Code, err.Error())
+		engine.logger.Errorf(time.Now(), im.GetInstanceID(), im.GetAttributes(), "Instance failed with error '%s': %s", cerr.Code, err.Error())
 	} else if errors.As(err, &uerr) && uerr.Code != "" {
 		engine.sugar.Errorf("Instance failed with uncatchable error '%s': %v", uerr.Code, err)
-		engine.logToInstance(ctx, time.Now(), im.cached, "Instance failed with uncatchable error '%s': %s", uerr.Code, err.Error())
+		engine.logger.Errorf(time.Now(), im.GetInstanceID(), im.GetAttributes(), "Instance failed with uncatchable error '%s': %s", uerr.Code, err.Error())
```
Comment (alankm) [inline pkg/flow/engine.go:406]: You've replaced the `CacheData` argument with two other arguments: `GetInstanceID` and `GetAttributes`. Why? Both of these things can be learned from the `CacheData`, can't they?

File: pkg/flow/states/parallel.go [hunk 2]
```diff
@@ -345,7 +345,7 @@ func (logic *parallelLogic) processActionResults(ctx context.Context, children [
 		children[idx].Complete = true
 		completed++
 
-		logic.Log(ctx, "Action returned. (%d/%d)", completed, len(children))
+		logic.SendToLogger(ctx, "info", "Action returned. (%d/%d)", completed, len(children))
```
Comment (alankm) [inline pkg/flow/states/parallel.go:348]: `Log` was roughly renamed to `SendToLogger`, but I don't think that was necessary. `Log` should have just been changed to support log levels. Or maybe another function should have been added to the interface alongside it: `LogError`.

File: pkg/flow/states/parallel.go [hunk 3]
```diff
@@ -329,7 +329,7 @@ func (logic *parallelLogic) processActionResults(ctx context.Context, children [
 
 		} else if results.ErrorMessage != "" {
 
-			logic.Log(ctx, "Branch %d crashed due to an internal error: %s", idx, results.ErrorMessage)
+			logic.SendToLogger(ctx, "error", "Branch %d crashed due to an internal error: %s", idx, results.ErrorMessage)
```
Comment (alankm) [inline pkg/flow/states/parallel.go:332]: All calls to `SendToLogger` are using strings to set log level. These should be constants.

**Pull request closed without merging**